File: backend/app/api/schemas/duty.py
"""Duty schemas."""
import datetime
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from ...models.enums import DutyType, Period

# A guard on the year field, not a business rule -- the same range
# `LeaveEntitlement`'s schemas use, for the same reason: it stops a typo'd
# year from creating a row that no duty range will ever resolve to.
MIN_DUTY_YEAR = 2000
MAX_DUTY_YEAR = 2100

# Matches DutyCounterAdjustment.adjustment's Numeric(5,1) column.
_ADJUSTMENT_LIMIT = 9999


class DutyBase(BaseModel):
    date: datetime.date
    period: Period
    doctor_id: int
    duty_type: DutyType


class DutyIn(DutyBase):
    """Create body -- the same fields as DutyBase, with no validation of
    its own.

    The day-of-week rule for secondary duty (it must land on the week's
    first *fully open* weekday -- neither AM nor PM closed -- which is only
    Monday when nothing is closed) needs PracticeClosure data, so it cannot
    be a stateless pydantic validator. It lives in routers/duty.py's
    create_duty, alongside the two other DB-dependent checks; see that
    module's docstring.
    """


class DutyOut(DutyBase):
    id: int
    model_config = {"from_attributes": True}

class DutyCountOut(BaseModel):
    """One doctor's duty count over the requested range, plus the
    adjustment the weighted score adds to it.

    `adjustment` is a `Decimal` and so serialises as a JSON string, like
    every other session quantity on this API (`DoctorOut.sessions_per_week`,
    the leave entitlement figures).
    """

    doctor_id: int
    doctor_code: str
    raw_count: int
    adjustment: Decimal


class DutyAdjustmentIn(BaseModel):
    """Upsert body for one doctor's duty counter adjustment in one year.

    Year-scoped because the count it adjusts is: the duty grid reads a whole
    calendar year, and by the next 1 January the count restarts and every
    doctor is genuinely level again.

    `target_count` is a *target effective total* for the year -- "make this
    doctor's duty counter read 14" -- not a credit. The endpoint stores
    `adjustment = target_count - raw_count` against the year's own count; see
    the identical shape in schemas/counter.py for why the derivation is
    server-side and why the field is not just called `count`.

    A target equal to the raw count derives a zero delta, which deletes the
    row rather than storing a zero, so the table holds only real deviations
    (the `LeaveEntitlement` principle).
    """

    doctor_id: int
    year: int = Field(ge=MIN_DUTY_YEAR, le=MAX_DUTY_YEAR)
    target_count: Decimal = Field(ge=-_ADJUSTMENT_LIMIT, le=_ADJUSTMENT_LIMIT)

    @model_validator(mode="after")
    def _one_decimal_place(self) -> "DutyAdjustmentIn":
        """Reject more than one decimal place, then normalise to exactly one --
        see the identical validator in schemas/counter.py."""
        quantized = self.target_count.quantize(Decimal("0.1"))
        if self.target_count != quantized:
            raise ValueError("target_count must have at most one decimal place")
        self.target_count = quantized
        return self

File: frontend/src/api/duty.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { apiClient } from "./client";
import type { DutyAssignment, DutyIn, DutyCount } from "./types";

export interface DutyCountsRange {
  from: string;
  to: string;
}

export const dutyKeys = {
  all: ["duty"] as const,
  list: () => [...dutyKeys.all, "list"] as const,
  counts: (range?: DutyCountsRange) =>
    range
      ? ([...dutyKeys.all, "counts", range.from, range.to] as const)
      : ([...dutyKeys.all, "counts"] as const),
};

/** Unfiltered - unlike leave, no doctor/date filter was called for. */
export function useDuty() {
  return useQuery({
    queryKey: dutyKeys.list(),
    queryFn: () => apiClient.get<DutyAssignment[]>("/duty"),
  });
}

export function useCreateDuty() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: DutyIn) => apiClient.post<DutyAssignment>("/duty", payload),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: dutyKeys.all });
    },
  });
}

export function useDeleteDuty() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => apiClient.delete<void>(`/duty/${id}`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: dutyKeys.all });
    },
  });
}

/**
 * When `range` is omitted, fetches unfiltered (all-time) counts. The
 * 4-weekly duty periods feature (see lib/date.ts) always passes a range
 * scoped to the selected period; the unranged form is kept only so any
 * other caller/test relying on the old behaviour is not silently changed.
 */
export function useDutyCounts(range?: DutyCountsRange) {
  return useQuery({
    queryKey: dutyKeys.counts(range),
    queryFn: () =>
      apiClient.get<DutyCount[]>(
        range ? `/duty/counts?from_date=${range.from}&to_date=${range.to}` : "/duty/counts",
      ),
  });
}
/**
 * Year-scoped, because the duty count it adjusts is: the grid reads a whole
 * calendar year, and by the next 1 January the count restarts and every
 * doctor is level again.
 *
 * `target_count` is the *effective total* the counter should read for that
 * year, not the adjustment: the server stores `target_count - raw_count`
 * derived against the 1 Jan-31 Dec count. The total sent must therefore be a
 * whole-year total - `GET /duty/counts` will happily count a narrower range,
 * and a total typed against one of those would store a delta the caller did
 * not mean. Sending a total equal to the raw count deletes the row
 * server-side, so the table holds only real deviations.
 *
 * Invalidates every duty key, not just the counts for this year: the counts
 * query key includes the range, and the adjustment applies to whichever range
 * starts in `year`.
 */
export interface DutyAdjustmentIn {
  doctor_id: number;
  year: number;
  target_count: string;
}

export function useSetDutyAdjustment() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: DutyAdjustmentIn) =>
      apiClient.put<DutyCount>("/duty/adjustment", payload),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: dutyKeys.all });
    },
  });
}
